libgen_parse_poc.py

```diff
@@ -0,0 +1,95 @@
+from typing import List, Dict
+import os
+import urllib.request
+from socket import timeout
+
+from tqdm import tqdm
+from libgen_api import Request, search, fetch
+from bs4 import BeautifulSoup
+
+
+# The library silently hangs (for minutes) if the mirror is unavailable.
+# TODO: patch it to have a reasonable timeout and raise a descriptive exception.
+fetch.LIBGEN_URL = 'https://libgen.is/search.php?'
+
+
+def filename(s: str) -> str:
+    return ''.join(c for c in s if c.isalnum() or c in "-_ .")
+
+
+class DomainAwareDownloader:
+    def __init__(self):
+        self.fails: Dict[str, int] = {}
+        self.trust_threshold = 2
+
+    @staticmethod
+    def domain(s: str) -> str:
+        return '/'.join(s.split('/')[:3])
+
+    def reorder_urls(self, urls: List[str]) -> List[str]:
+        urls_trusteds = [
+                (url, bool(self.fails.get(self.domain(url), 0) < self.trust_threshold))
+                for url in urls
+            ]
+        trusted_urls = [url for url, is_trusted in urls_trusteds if is_trusted]
+        untrusted_urls = [url for url, is_trusted in urls_trusteds if not is_trusted]
+        return trusted_urls + untrusted_urls
+
+    def try_download(self, name: str, urls: List[str]) -> None:
+        # A-la `mkdir -p`
+        try:
+            os.mkdir('books')
+        except FileExistsError:
+            pass
+
+        urls = self.reorder_urls(urls)
+
+        data = None
+        for url in urls:
+            try:
+                with urllib.request.urlopen(url, timeout=5) as conn:
+                    data = conn.read()
+            except urllib.error.URLError as e:
+                # I only want to catch actual timeout errors here. But can't specify
+                # that in the `except` because `urllib` wraps the actual error in its
+                # own type...
+                if not isinstance(e.reason, TimeoutError):
+                    raise
```

Comment:
is it possible to skip this url if it fails, and continue the parsing? I've got several broken links which ruins all procedure

Reply:
That's what it was supposed to do (the exception should've been `urllib.error.URLError` with `e.reason` being `TimeoutError`). If that's not what happened (i.e. you received an error in the output and the script stopped) then it's a bug. Please, share the traceback and I will attempt improve the error handling.

I hate to do something like a bare `except`, as it would silently suppress any errors and a user would get no clue what is going on. For instance, it might be useful to distinguish between an unreachable mirror and a url that was reached but the download timed out (which may be due to the book file being too large)

Reply:
For example, url `'http://62.182.86.140/main/1000/b14e74047c58550bc36b25c79fba24ff/Graham%20Hutton%20-%20Programming%20in%20Haskell-Cambridge%20University%20Press%20%282007%29.pdf'` throws an `urllib.error.URLError: <urlopen error timed out> `error. Can we skip this url and try another mirror?

Reply:
Got it. `TimeoutError` and `socket.timeout` became the same thing since python 3.10 but were two distinct exception classes in 3.9 and earlier. Will fix in a moment